dev/core#3961 Create stub extensions for all components

colemanw opened the pull request:
Overview
----------------------------------------
Implements [dev/core#3961 proposal to create a stub extension for all components](https://lab.civicrm.org/dev/core/-/issues/3961).

Before
----------------------------------------
There are 8 components in CiviCRM (_CiviMail_, _CiviEvent_, _CiviContribute_, etc)

After
----------------------------------------
Now there are also 8 new extensions (`civi_mail`, `civi_event`, `civi_contribute`, etc). The extensions do nothing (yet), but are enabled/disabled in-sync with their corresponding components.

Technical Details
----------------------------------------
- **Naming convention:** Is a simple conversion from `CamelCase` to `snake_case`. (e.g. `CiviCase` component == `civi_case` extension). This is easy to convert, doesn't require any extra mapping, and avoids conflicts with other extensions in the civi-universe, (namely the `civicase` extension which would conflict if we didn't use an underscore in our name).
- **Status-sync:** 2-way syncing: enabling the extension will trigger the component to be enabled, and vice-versa.
- **Upgrader:** This uses the [2nd variant](https://lab.civicrm.org/dev/core/-/issues/3961#note_84070) suggested by @totten, which is a module-extension with shared upgrader class to manage the enable/disable logic.

Next Steps
-------------------
This PR is a baby step to add the stub extensions with no functionality, just so we can work the kinks out of the install-sync mechanisms. Once that's done and this PR is merged, a lot of possibilities open up such as:
- Getting rid of the *Enable Components* screen
- Packaging SearchKit searches with components
- Migrating code out of core and into the extensions

colemanw committed:
Add stub extensions for all components

colemanw committed:
Sync enable/disable status of component extensions with core components

colemanw committed:
Add test for component-extension lifecycle

totten commented:
@colemanw That looks pretty. :)

🌹 

Regarding the error, it seems fairly reproducible for me. My steps for reproducing:

```bash
#### Setup
civibuild reinstall dmaster
## ... wait... build script fails updating `enable_components`
civibuild snapshot dmaster

#### Run (and re-run and re-run...)
civibuild restore dmaster
cv vset '{"enable_components":["CiviEvent","CiviContribute","CiviMember","CiviMail","CiviReport","CiviPledge","CiviCase","CiviCampaign"]}'
```

The verbose output from `cv vset -vv ...` gives a backtrace:

```
In Error.php line 954:

  [Civi\Core\Exception\DBQueryException]
  DB Error: already exists


Exception trace:
  at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/CRM/Core/Error.php:954
 CRM_Core_Error::exceptionHandler() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:945
 PEAR_Error->__construct() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/db/DB.php:997
 DB_Error->__construct() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:575
 PEAR::_raiseError() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:223
 PEAR->__call() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/db/DB/common.php:1928
 DB_common->raiseError() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/db/DB/mysqli.php:943
 DB_mysqli->mysqliRaiseError() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/db/DB/mysqli.php:413
 DB_mysqli->simpleQuery() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/vendor/pear/db/DB/common.php:1234
 DB_common->query() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/packages/DB/DataObject.php:2696
 DB_DataObject->_query() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/packages/DB/DataObject.php:1245
 DB_DataObject->insert() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/CRM/Extension/Manager.php:669
 CRM_Extension_Manager->_createExtensionEntry() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/CRM/Extension/Manager.php:297
 CRM_Extension_Manager->install() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/CRM/Core/Component.php:454
 CRM_Core_Component::onToggleComponents() at n/a:n/a
 call_user_func() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/Civi/Core/SettingsBag.php:439
 Civi\Core\SettingsBag->setDb() at /Users/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/Civi/Core/SettingsBag.php:250
 Civi\Core\SettingsBag->set() at phar:///Users/totten/bknix/bin/cv/src/Command/SettingSetCommand.php:47
 Civi\Cv\Command\SettingSetCommand->execute() at phar:///Users/totten/bknix/bin/cv/vendor/symfony/console/Command/Command.php:127
 Cvphar\Symfony\Component\Console\Command\Command->run() at phar:///Users/totten/bknix/bin/cv/vendor/symfony/console/Application.php:637
 Cvphar\Symfony\Component\Console\Application->doRunCommand() at phar:///Users/totten/bknix/bin/cv/vendor/symfony/console/Application.php:190
 Cvphar\Symfony\Component\Console\Application->doRun() at phar:///Users/totten/bknix/bin/cv/src/Application.php:66
 Civi\Cv\Application->doRun() at phar:///Users/totten/bknix/bin/cv/vendor/symfony/console/Application.php:101
 Cvphar\Symfony\Component\Console\Application->run() at phar:///Users/totten/bknix/bin/cv/src/Application.php:32
 Civi\Cv\Application::main() at phar:///Users/totten/bknix/bin/cv/bin/cv:28
 require() at /Users/totten/bknix/bin/cv:14
```

One idea:

* It's a bi-directional sync: ___(1=>2)___ from `civicrm_setting` to `civicrm_extension`; and also ___(2=>1)___ from `civicrm_extension` to `civicrm_setting`). But they shouldn't really fire each other (__1=>2=>1=>2...__)
* In my first read of `CRM_Extension_Upgrader_Component` and `CRM_Core_Component::onToggleComponents()`, I don't recognize any conditionals that would prevent them from firing each other. (*That doesn't mean there's nothing there -- just nothing obvious to me.*)
* If there was some kind of loop that caused it to insert `civicrm_extension` multiple times, then the error `DB Error: already exists` would make sense.

totten commented:
@colemanw Here's a visualization of how they call each other:

```
>> ["SettingsBag::setDb(enable_components, [\"CiviEvent\",\"CiviContribute\",\"CiviMember\",\"CiviMail\",\"CiviReport\",\"CiviPledge\",\"CiviCase\",\"CiviCampaign\"] )","(via=Civi\\Core\\SettingsBag::set)"]
   >> ["SettingsBag::setDb(resCacheCode, \"X7mcf\" )","(via=Civi\\Core\\SettingsBag::set)"]
   >> ["(via=::call_user_func)"]
      >> ["(via=CRM_Core_Component::onToggleComponents)"]
         >> ["Install civi_case","(via=CRM_Extension_Upgrader_Base::onInstall)"]
            >> ["enableComponent(CiviCase)","(via=CRM_Extension_Upgrader_Component::install)"]
               >> ["SettingsBag::setDb(enable_components, [\"CiviEvent\",\"CiviContribute\",\"CiviMember\",\"CiviMail\",\"CiviReport\",\"CiviPledge\",\"CiviCase\"] )","(via=Civi\\Core\\SettingsBag::set)"]
                  >> ["SettingsBag::setDb(resCacheCode, \"nQ2F9\" )","(via=Civi\\Core\\SettingsBag::set)"]
                  >> ["(via=::call_user_func)"]
                     >> ["(via=CRM_Core_Component::onToggleComponents)"]
                        >> ["Install civi_case","(via=CRM_Extension_Upgrader_Base::onInstall)"]
                           >> ["enableComponent(CiviCase)","(via=CRM_Extension_Upgrader_Component::install)"]
                              >> ["SettingsBag::setDb(enable_components, [\"CiviEvent\",\"CiviContribute\",\"CiviMember\",\"CiviMail\",\"CiviReport\",\"CiviPledge\",\"CiviCase\"] )","(via=Civi\\Core\\SettingsBag::set)"]
                                 >> ["SettingsBag::setDb(resCacheCode, \"aFAC5\" )","(via=Civi\\Core\\SettingsBag::set)"]
                        >> ["Enable civi_case","(via=CRM_Extension_Upgrader_Base::onEnable)"]
                           >> ["enableComponent(CiviCase)","(via=CRM_Extension_Upgrader_Component::enable)"]
                        >> ["_createExtensionEntry(civi_case)","(via=CRM_Extension_Manager::install)"]
                        >> ["SettingsBag::setDb(resCacheCode, \"OC0RL\" )","(via=Civi\\Core\\SettingsBag::set)"]
         >> ["Enable civi_case","(via=CRM_Extension_Upgrader_Base::onEnable)"]
            >> ["enableComponent(CiviCase)","(via=CRM_Extension_Upgrader_Component::enable)"]
         >> ["_createExtensionEntry(civi_case)","(via=CRM_Extension_Manager::install)"]
```

Generated with https://gist.github.com/totten/fcf07d85b393411584b8f60ba273292a

This does seem to show a recursive loop between the calls for

* `SettingsBag::setDb(enable_components,...)`
* `"Install civi_case","(via=CRM_Extension_Upgrader_Base::onInstall)"`
* `"SettingsBag::setDb(enable_components, ...)`
* `"Install civi_case","(via=CRM_Extension_Upgrader_Base::onInstall)"`

Each loop has a call to `_createExtensionEntry()`. The second call to "_createExtensionEntry(civi_case)" causes the DB exception.

colemanw committed:
Component/Extensions - Remove unnecessary cache clearing and prevent recursion

Ensure the cached settings are updated _before_ post_change hooks are called for settings.

All the cache clears being performed by on_change hooks were redundant with what happens when
extensions are enabled/disabled so I removed it.

Also moved the extension install hook to use postInstall which should be more stable, and
added checks in onToggleComponents to prevent recursion.

colemanw commented:
Thanks for the analysis @totten. According to the unit test I added, there ought to be no recursion, and there are safeguards against recursion on both sides of the equation:

- ___(1=>2)___ from `civicrm_setting` to `civicrm_extension`: 
    1. It checks to see if the setting is actually being changed before enabling/disabling any extensions.
    2. Also, calling `enable()` on an enabled extension (or `disable()` on a disabled one) won't trigger the callbacks in `CRM_Extension_Upgrader_Component`. 
- ___(2=>1)___ from `civicrm_extension` to `civicrm_setting`:
    1. The `CRM_Core_BAO_ConfigSetting::enable/disableComponent` functions both have early returns if the component already has that status. This should short-circuit any recursion.

The unit test I added passes locally, yet despite all this, I must be missing something because the build still fails.
I thought I was on to something when I discovered that a local copy of the extension manager can cause problems, but even after I removed all code like `$manager = CRM_Extension_System::singleton()->getManager();` in favor of calling `getManager()` for a fresh copy every time, it still didn't solve the problem so I spun a side-issue for that #26113.

colemanw commented:
Ok @totten following your steps to reproduce I was able to pinpoint the problem, it was indeed a caching issue. I had to ensure that cached settings are updated _before_ post_change hooks are called.
retest this please

colemanw committed:
Component Extensions - Fix upgrade and test errors

colemanw commented:
@totten we are down to a single test failure (the same problem in several upgrader tests). I've addressed that with #26118, and then this PR I think should wait to be merged next week after branching so we can wait for the domain/settings stuff to settle.

totten commented:
I've had a problem in this scenario:

1. __Setup__: Install 5.60. Toggle some components (CiviPledge=>off, CiviCase=>off)
2. __Upgrading__: Checkout master+PR. Run `composer install`. Run `cv upgrade:db`
    * Confirm we have same menu/components as pre-upgrade. List matches in both "Admin>Components" and "Admin>Extensions". Huzzah.
3. __Disabling CiviCampaign__:
    * In "Admin>Components", disable "CiviCampaign".
        * Confirm main menu is updated. Huzzah.
    * In "Admin>Extensions", check "CiviCampaign". (Similarly, in CLI, check `cv ext:list -Li`)
        * Uh oh! It still shows CiviCampaign as enabled.
4. __Enabling CiviCase__:
    * In "Admin>Components", enable "CiviCase".
        * Confirm main menu is updated. Huzzah.
    * In "Admi>Extensions", check "CiviCase". (Similarly, in CLI, check `cv ext:list -Li`)
        * It's been added. Huzzah.

I'm not sure why step 3 (Disabling CiviCampaign) had trouble.

However, on the plus side... after continuing to 4 (Enabling CiviCase), I did a few more rounds of enable/disable - and they seemed to work fine. I played around with using both "Admin>Components" and "Admin>Extensions" to enable/disable things, and it seemed to become pretty stable.  

The original sequence (1=>2=>3=>4) seems to reliably have the same problem during step 3. However, if I switch the order (1=>2=>4=>3), then it's fine. Feels like maybe some leftover inaccuracy in metadata?

totten commented:
On the theory of some cache-style issue, I tried interjecting a step:

* (1 => 2 => `cv flush` => 3) It doesn't matter if I call `cv flush`. Step 3 still has the same problem.
* (1 => 2 => Open/Save Components => 4) If I open "Admin>Components" and save the screen (*without any changes*), then this does make a difference. Subsequent updates to "Admin>Components" will propagate as expected.

totten commented:
@colemanw FWIW, I can confirm that this fixed the `r-run` scenarios from above. :) Alas, a new test-failure. :(

colemanw commented:
retest this please

colemanw commented:
@totten looks like the unit test wasn't taking into account the possibility that disabled extensions might exist in the test database. Fixed with d831b8ca67b0aa4cc918121fde6beebd0b846839
If you want to give this `merge-ready` I'll rebase to clean up the commit history then merge.

totten commented:
Yaaaaay. Tests pass. I re-ran the same procedures as before and tried some additional ones (e.g. using CLI to configure via setting and/or module) - all seemed good. 🎉
